src/ucp/proto/proto_select.h

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCP_PROTO_SELECT_H_
+#define UCP_PROTO_SELECT_H_
+
+#include "proto.h"
+
+#include <ucs/datastruct/khash.h>
+
+
+/**
+ * Some flags from ucp_request_param_t.op_attr_mask can affect protocol
+ * selection decision.
+ */
+#define UCP_PROTO_SELECT_OP_ATTR_BASE   UCP_OP_ATTR_FLAG_NO_IMM_CMPL
+#define UCP_PROTO_SELECT_OP_ATTR_MASK   UCP_OP_ATTR_FLAG_FAST_CMPL
+
+
+/**
+ * Entry which defines which protocol should be used for a message size range.
+ */
+typedef struct {
+    ucp_proto_config_t          proto_config;   /* Protocol configuration to use */
+    size_t                      max_msg_length; /* Max message length, inclusive */
+} ucp_proto_threshold_elem_t;
+
+
+/**
+ * Protocol selection per a particular buffer type and operation
+ */
+typedef struct {
+    ucp_proto_threshold_elem_t  *thresholds; /* Array of which protocol to use
```

Comment:
do we need the size of this array? How it would be used?

Reply:
the last element in the array has SIZE_MAX as max_msg_length